src/Features/Core/Portable/ExternalAccess/UnitTesting/API/UnitTestingHotReloadService.cs

```diff
@@ -0,0 +1,124 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.Api
+{
+    internal sealed class UnitTestingHotReloadService
+    {
+        private sealed class DebuggerService : IManagedEditAndContinueDebuggerService
+        {
+            public static readonly DebuggerService Instance = new();
+
+            public Task<ImmutableArray<ManagedActiveStatementDebugInfo>> GetActiveStatementsAsync(CancellationToken cancellationToken)
+                => Task.FromResult(ImmutableArray<ManagedActiveStatementDebugInfo>.Empty);
+
+            public Task<ManagedEditAndContinueAvailability> GetAvailabilityAsync(Guid module, CancellationToken cancellationToken)
+                => Task.FromResult(new ManagedEditAndContinueAvailability(ManagedEditAndContinueAvailabilityStatus.Available));
+
+            // TODO: get capabilities from the runtime
+            public Task<ImmutableArray<string>> GetCapabilitiesAsync(CancellationToken cancellationToken)
+                => Task.FromResult(ImmutableArray.Create("Baseline", "AddDefinitionToExistingType", "NewTypeDefinition"));
+
+            public Task PrepareModuleForUpdateAsync(Guid module, CancellationToken cancellationToken)
+                => Task.CompletedTask;
+        }
+
+        public readonly struct Update
+        {
+            public readonly Guid ModuleId;
+            public readonly ImmutableArray<byte> ILDelta;
+            public readonly ImmutableArray<byte> MetadataDelta;
+            public readonly ImmutableArray<byte> PdbDelta;
+
+            public Update(Guid moduleId, ImmutableArray<byte> ilDelta, ImmutableArray<byte> metadataDelta, ImmutableArray<byte> pdbDelta)
+            {
+                ModuleId = moduleId;
+                ILDelta = ilDelta;
+                MetadataDelta = metadataDelta;
+                PdbDelta = pdbDelta;
+            }
+        }
+
+        private static readonly ActiveStatementSpanProvider s_solutionActiveStatementSpanProvider =
+            (_, _, _) => ValueTaskFactory.FromResult(ImmutableArray<ActiveStatementSpan>.Empty);
+
+        private readonly IEditAndContinueWorkspaceService _encService;
+        private DebuggingSessionId _sessionId;
+
+        public UnitTestingHotReloadService(HostWorkspaceServices services)
+            => _encService = services.GetRequiredService<IEditAndContinueWorkspaceService>();
+
+        /// <summary>
+        /// Starts the watcher.
+        /// </summary>
+        /// <param name="solution">Solution that represents sources that match the built binaries on disk.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public async Task StartSessionAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var newSessionId = await _encService.StartDebuggingSessionAsync(solution, DebuggerService.Instance, captureMatchingDocuments: true, reportDiagnostics: false, cancellationToken).ConfigureAwait(false);
```

Comment:
Perhaps we should pass in a string here, for telemetry purposes, to track where the session came from. "Unit testing", "Watch", "IDE" etc.

Reply:
Yes, eventually we'll do something like that. https://github.com/dotnet/roslyn/issues/52128

Right now we don't report telemetry for Hot Reload to avoid tainting EnC telemetry.